### DISCUSSION TITLE
Local energy minimization question

### DISCUSSION THREAD

**Discussion by haroldgrosjean** [rdkit/rdkit#4528] [Q&A] [answered]
Hello,

For a QM simulation I have to extract a protein-ligand complex binding site residues. This will create breaks in the backbone with incorrect hybridization states. The way I am trying to correct that is by adding neighboring backbone atoms to the atom group (in MDanalysis) and generate a PDB from that:

[F126_BS_MOL_mda.pdb.txt](https://github.com/rdkit/rdkit/files/7179193/F126_BS_MOL_mda.pdb.txt)

Then, I plug this file into rdkit to change the nitrogens into carbons, modify hybridization states of these "fake" carbons to from SP2 to SP3 and add hydrogens to these atoms.

Here is the code:
```  
import rdkit as rdk
from rdkit import Chem
from rdkit.Chem import AllChem

F126_BS_mol_rdk = rdk.Chem.MolFromPDBFile('F126_BS_MOL_mda.pdb.txt', removeHs=False)
#This list was defined in MDAnalysis prior to the production of the above PDB
atoms_to_H = [384, 385, 386, 387, 388, 389, 390, 391, 392, 393, 394, 395, 396, 397, 398, 399, 400, 401, 402, 403]
for atom in F126_BS_mol_rdk.GetAtoms():
    if atom.GetAtomicNum() == 7 and atom.GetIdx() in atoms_to_H::
        atom.SetAtomicNum(6)
        atom.SetHybridization(Chem.rdchem.HybridizationType.SP3)
        atom.SetNumExplicitHs(1)
F126_BS_mol_prot_rdk = Chem.AddHs(F126_BS_mol_rdk, addCoords=True, onlyOnAtoms=atoms_to_H)
Chem.MolToPDBFile(F126_BS_mol_prot_rdk, './F126_BS_H.pdb')
```
the molecule produced by this code is almost perfect but there one problem that I don't know how to address. The hydrogens seems to have been added in a semi-random way with non-optimal geometries which would create clashes during my calculation. See the images at the end. I tried to minimize the whole molecule with with:
```
AllChem.MMFFOptimizeMoleculeConfs(F126_BS_mol_prot_rdk, numThreads=12)
```
But this also distorts the rest of the correctly positioned atoms. Could you think of way to perform energy minimization on just the atoms listed in ```atoms_to_H``` and bonds with neighboring atoms whilst keeping the rest of the atoms frozen?

many thanks in advance for your help,

Regards,

Harold Grosjean

![Screenshot from 2021-09-16 17-13-27](https://user-images.githubusercontent.com/51821930/133647967-dfef4c64-b0b3-4814-b7c0-bf0e527aae80.png)

![Screenshot from 2021-09-16 17-16-24](https://user-images.githubusercontent.com/51821930/133648677-4d292719-4444-4932-9a77-41cb1997ca8f.png)

**Comment by ptosco** [comment, marked as answer]
Hi Harold,

you can set positional restraints on all heavy atoms. Unfortunately, currently this does not play well with `OptimizeMoleculeConfs` as the reference coordinates are set based on the conformation that you construct the forcefield with, while each conformation would need its own reference coordinates. I'll look into improving that.
Meanwhile, you can use `multiprocessing` to achieve parallel minimization of your conformations keeping all heavy atoms restrained to their original positions; see an example below that you should manage to adapt to your own molecules:

```
from rdkit import Chem
from rdkit.Chem import AllChem
import multiprocessing

mol = Chem.AddHs(Chem.MolFromSmiles("O=C(O)C(c1ccc(cc1)CCN4CCC(c2nc3ccccc3n2CCOCC)CC4)(C)C"))
cids = AllChem.EmbedMultipleConfs(mol)

with Chem.SDWriter("orig_confs.sdf") as w:
    for cid in cids:
        w.write(mol, confId=cid)

mp = AllChem.MMFFGetMoleculeProperties(mol)

def minimize_thread(mol):
    ff = AllChem.MMFFGetMoleculeForceField(mol, mp)
    for a in mol.GetAtoms():
        if (a.GetAtomicNum() > 1):
            ff.MMFFAddPositionConstraint(a.GetIdx(), 0.0, 1.e4)
    ff.Minimize(maxIts=10000)
    return mol

pool = multiprocessing.Pool(12)

# turn the multi-conf mol into an array of mols with 1 conformation each
single_conf_mols = []
for cid in cids:
    mol_copy = Chem.Mol(mol, True)
    mol_copy.AddConformer(mol.GetConformer(cid))
    single_conf_mols.append(mol_copy)

# minimize using multiple CPUs
min_mols = pool.map(minimize_thread, single_conf_mols)

# rebuild a multi-conf minimized mol
min_mol = Chem.Mol(mol, True)
for mm in min_mols:
    min_mol.AddConformer(mm.GetConformer())

with Chem.SDWriter("min_confs.sdf") as w:
    for cid in cids:
        w.write(min_mol, confId=cid)
```

**Reply by haroldgrosjean** [reply]
Great that worked! Thanks a lot!

**Reply by fdaqin** [reply]
@ptosco @haroldgrosjean Hi, ptosco, harold, I have a related problem here. Currently, we extract the ligand and pocket residues for optimization in rdkit and use similar codes. However, a large number of clashes were generated after optimization, between ligand and pocket residues, and between residues. Why? Since the pocket is also composed of discontinuous residues, so this mol object actually contains many independent molecules. Is it because rdkit does not consider the intermolecular interactions that lead to the generation of clashes? Thanks!

**Reply by haroldgrosjean** [reply]
Hello! Are you sure that you're adding the constraint to the heavy atoms correctly? Because the above worked well in my case and it seems that the problem is similar. Maybe writing your code would help us troubleshoot your problem?

Bests,

Harold

**Reply by ptosco** [reply]
@fdaqin Based on your decription of the problem I suspect that this depends on how you are constructing your force field:
http://rdkit.org/docs/source/rdkit.Chem.rdForceFieldHelpers.html?highlight=forcefield#rdkit.Chem.rdForceFieldHelpers.MMFFGetMoleculeForceField
```
rdkit.Chem.rdForceFieldHelpers.MMFFGetMoleculeForceField((Mol)mol,
(MMFFMolProperties)pyMMFFMolProperties[, (float)nonBondedThresh=100.0
[, (int)confId=-1[, (bool)ignoreInterfragInteractions=True]]]) → ForceField
```
By default `ignoreInterfragInteractions` is set to `True`, which would explain the clashes between independent molecules. If you set it to `False` I expect the clashes to disappear. `nonBondedThreshold` is most likely fine since it defaults to 100.0 angstrom.

**Reply by fdaqin** [reply]
@ptosco @haroldgrosjean Thanks for your kind reply! And ptosco, you are right, it works!
 In the following figures, the pocket residues were all constrained, only the ligand was optimized. When using default settings, there will be many clashes between ligand and residues after minimization.
![figure1](https://user-images.githubusercontent.com/33869719/212340591-b927136e-b70e-4d8f-ac37-9a0915d4c7ab.png)

When ignoreInterfragInteractions was set to False, there is no clash between ligand and pocket residues 
![figure2](https://user-images.githubusercontent.com/33869719/212340628-288f2167-ff6e-4a7d-98d4-65afdf19aa9a.png)
